### PR TITLE
Add new notification for Root Prefab Instance's First Load

### PR DESCRIPTION
**Pull request by chiyenteng**
Add new notification `OnRootPrefabInstanceLoaded` to inform users when the first root prefab instance load is done after loading current level file.

Tested manually with level opening/creating/switching. The breakpoint on code for sending `OnRootPrefabInstanceLoaded` is always triggered once the level being loaded, and all components in the root prefab instance's level container entity is de-serialized correctly when firing the notification.

Signed-off-by: chiyenteng <82238204+chiyenteng@users.noreply.github.com>

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h [hunk 1]
```diff
@@ -30,6 +30,14 @@ namespace AzToolsFramework
 
             // Remove an Instance from the waiting queue.
             virtual void RemoveTemplateInstanceFromQueue(const Instance* instance) = 0;
+
+            // Set the executor as it never loads root prefab instance.
+            // A notification OnRootPrefabInstanceFirstLoadEnd will fire if root prefab instance
+            // is loaded for the first time after this function is called.
+            virtual void SetRootPrefabInstanceAsNeverLoaded() = 0;
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h:37]: Hmm, I think something like `QueueRootPrefabLoadedNotificationForNextPropagation` might be a clearer indication of intention here, thoughts?

**Review by NicholasVanSickle — COMMENTED** [review]
Logic looks good, this will be a super useful notification to have (though we'll eventually have to disambiguate root prefabs, I'm OK with not tackling this now). I have a few nits on naming, but we're otherwise good to go.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h (hunk 1 above)
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h]: Renamed.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h [hunk 2]
```diff
@@ -30,6 +30,14 @@ namespace AzToolsFramework
...
+
+            // Check if root prefab instance's first load is ended after loading level file.
+            virtual bool IsRootPrefabInstanceFirstLoadEnded() const = 0;
```
[the remark this answers is not in the document]
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Instance/InstanceUpdateExecutorInterface.h]: Removed.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicNotificationBus.h [hunk 3]
```diff
@@ -23,6 +23,7 @@ namespace AzToolsFramework
 
             virtual void OnPrefabInstancePropagationBegin() {}
             virtual void OnPrefabInstancePropagationEnd() {}
+            virtual void OnRootPrefabInstanceFirstLoadEnd() {}
```
[the remark this answers is not in the document]
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicNotificationBus.h]: Renamed.

**Review by NicholasVanSickle — APPROVED** [review]

**Review by hultonha — COMMENTED** [review]
Thank you very much for adding this new notification @chiyenteng and the change itself looks good! I do really think this should come with an accompanying test to validate and encode the behavior to help ensure it's doing what we think it's doing today, and guard against any future regressions (edits/maintenance updates etc...)

I think this could be in the form of an integration test in `AzToolsFramework` Test to ensure the notification is received at the right time (e.g. A particular component can be accessed that couldn't be before in `OnEntityStreamLoadSuccess`). I'm sure we have something similar already and it shouldn't be too difficult to add. I'm happy to try and help with this too and potentially @srikappa-amzn might have an idea too. Thanks!

**Review by srikappa-amzn — APPROVED** [review]

**Comment by srikappa-amzn** [comment]
> Thank you very much for adding this new notification @chiyenteng and the change itself looks good! I do really think this should come with an accompanying test to validate and encode the behavior to help ensure it's doing what we think it's doing today, and guard against any future regressions (edits/maintenance updates etc...)
> 
> I think this could be in the form of an integration test in `AzToolsFramework` Test to ensure the notification is received at the right time (e.g. A particular component can be accessed that couldn't be before in `OnEntityStreamLoadSuccess`). I'm sure we have something similar already and it shouldn't be too difficult to add. I'm happy to try and help with this too and potentially @srikappa-amzn might have an idea too. Thanks!

We don't have current tests in AzToolsFramework around level loading. A fixture needs to be setup and call PrefabEditorEOS. We can either build a quick json for the level in the test itself or do something like 
```
        AZStd::vector<char> charBuffer;
        AZ::IO::ByteContainerStream<AZStd::vector<char>> stream(&charBuffer);
        AZ::Utils::SaveObjectToStream<AZ::Entity>(stream, AZ::ObjectStream::ST_XML, levelObject, m_app->GetSerializeContext());
        stream.Seek(0, AZ::IO::GenericStream::ST_SEEK_BEGIN);
```
to build the stream and pass it to LoadFromStream().

It's a bit more involved to get these tests up and running. I suggest adding them as a follow up action item and allow these changes to go to development. @chiyenteng , please add a follow up task for the tests and get them into development soon.

**Comment by chiyenteng** [comment]
> It's a bit more involved to get these tests up and running. I suggest adding them as a follow up action item and allow these changes to go to development. @chiyenteng , please add a follow up task for the tests and get them into development soon.

Will do.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Entity/PrefabEditorEntityOwnershipService.cpp [hunk 4]
```diff
@@ -233,6 +234,8 @@ namespace AzToolsFramework
         m_rootInstance->SetTemplateId(templateId);
         m_rootInstance->SetTemplateSourcePath(m_loaderInterface->GenerateRelativePath(filename));
         m_rootInstance->SetContainerEntityName("Level");
+
+        AZ::Interface<Prefab::InstanceUpdateExecutorInterface>::Get()->QueueRootPrefabLoadedNotificationForNextPropagation();
```
[the remark this answers is not in the document]
Comment (chiyenteng) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Entity/PrefabEditorEntityOwnershipService.cpp]: Validation added.

**Comment by hultonha** [comment]
> > Thank you very much for adding this new notification @chiyenteng and the change itself looks good! I do really think this should come with an accompanying test to validate and encode the behavior to help ensure it's doing what we think it's doing today, and guard against any future regressions (edits/maintenance updates etc...)
> > I think this could be in the form of an integration test in `AzToolsFramework` Test to ensure the notification is received at the right time (e.g. A particular component can be accessed that couldn't be before in `OnEntityStreamLoadSuccess`). I'm sure we have something similar already and it shouldn't be too difficult to add. I'm happy to try and help with this too and potentially @srikappa-amzn might have an idea too. Thanks!
> 
> We don't have current tests in AzToolsFramework around level loading. A fixture needs to be setup and call PrefabEditorEOS. We can either build a quick json for the level in the test itself or do something like
> 
> ```
>         AZStd::vector<char> charBuffer;
>         AZ::IO::ByteContainerStream<AZStd::vector<char>> stream(&charBuffer);
>         AZ::Utils::SaveObjectToStream<AZ::Entity>(stream, AZ::ObjectStream::ST_XML, levelObject, m_app->GetSerializeContext());
>         stream.Seek(0, AZ::IO::GenericStream::ST_SEEK_BEGIN);
> ```
> 
> to build the stream and pass it to LoadFromStream().
> 
> It's a bit more involved to get these tests up and running. I suggest adding them as a follow up action item and allow these changes to go to development. @chiyenteng , please add a follow up task for the tests and get them into development soon.

Thanks for the info @srikappa-amzn and thanks for creating the follow-up task @chiyenteng. Do you know roughly when this will be implemented as this sort of thing is super important for the long-term success of Prefabs. It's worth investing the time now to help guard against future regressions and catch failures early.

Tagging @monroegm for visibility, hopefully https://github.com/o3de/o3de/issues/9662 can be prioritized for the next sprint potentially. This will then also help with future changes once the test infrastructure is in place. Thanks!

**Pull request merged**